localstack/services/sns/provider.py

```diff
@@ -713,62 +571,104 @@ def publish(
 
         if message_attributes:
             validate_message_attributes(message_attributes)
 
         store = self.get_store()
 
-        # No need to create a topic to send SMS or single push notifications with SNS
-        # but we can't mock a sending so we only return that it went well
-        if not phone_number and not target_arn:
-            if topic_arn not in store.sns_subscriptions:
-                raise NotFoundException(
-                    "Topic does not exist",
-                )
-        # Legacy format to easily leverage existing publishing code
-        # added parameters parsed by ASF. TODO: check/remove
-        req_data = {
-            "Action": ["Publish"],
-            "TopicArn": [topic_arn],
-            "TargetArn": target_arn,
-            "Message": [message],
-            "MessageAttributes": [message_attributes],
-            "MessageDeduplicationId": [message_deduplication_id],
-            "MessageGroupId": [message_group_id],
-            "MessageStructure": [message_structure],
-            "PhoneNumber": [phone_number],
-            "Subject": [subject],
-        }
-        message_id = publish_message(
-            topic_arn, req_data, context.request.headers, message_attributes=message_attributes
+        if not phone_number:
+            if is_endpoint_publish:
+                moto_sns_backend = sns_backends[context.account_id][context.region]
+                if target_arn not in moto_sns_backend.platform_endpoints:
+                    raise InvalidParameterException(
+                        "Invalid parameter: TargetArn Reason: No endpoint found for the target arn specified"
+                    )
+            else:
+                topic = topic_arn or target_arn
+                if topic not in store.sns_subscriptions:
+                    raise NotFoundException(
+                        "Topic does not exist",
+                    )
+
+        message_ctx = SnsMessage(
+            type="Notification",
+            message=message,
+            message_attributes=message_attributes,
+            message_deduplication_id=message_deduplication_id,
+            message_group_id=message_group_id,
+            message_structure=message_structure,
+            subject=subject,
         )
-        return PublishResponse(MessageId=message_id)
+        publish_ctx = SnsPublishContext(
+            message=message_ctx, store=store, request_headers=context.request.headers
+        )
+
+        if is_endpoint_publish:
+            self._publisher.publish_to_application_endpoint(
+                ctx=publish_ctx, endpoint_arn=target_arn
+            )
+        elif phone_number:
+            self._publisher.publish_to_phone_number(ctx=publish_ctx, phone_number=phone_number)
+        else:
+            # TODO: beware if FIFO, order is guaranteed yet. Semaphore? might block workers
+            # 2 quick call in succession might be unordered in the executor? need to try it with many threads
```

Comment:
Yes, I've tried to rewrite it a bit, and I'll explain it here so maybe you can tell me if you see what I mean 😅
New todo: 
```python
# TODO: beware if the subscription is FIFO, the order might not be guaranteed.
# 2 quick call to this method in succession might not be executed in order in the executor?
```

Now, we're using a `ThreadPoolExecutor` to actually publish, we submit tasks to workers.
The issue I thought is that we send 2 calls to LocalStack very close to each other, it's not guaranteed that the first call will execute first, thus not sending the message in a FIFO manner (one worker might execute faster, I'm not sure exactly how it would happen but I guess race conditions are always tricky).
One idea was to use some kind of synchronisation in case of FIFO queue, saving the order of received calls to actually send those message in the right order, but this might block workers, well, it's multithreaded synchronisation, which is always tricky.
I just wrote a todo to remember that something might happen, but the executor should submit and execute pretty sequentially, so we should be fine.

Reply:
I see what you mean now.
nit: Now it is not really clear anymore what the actual todo is^^.
Suggestion:
```
# beware if the subscription is FIFO, the order might not be guaranteed.
#  2 quick calls to this method in succession might not be executed in order in the executor.
# TODO: test how this behaves in a FIFO context with a lot of threads.
```

Reply:
Done 👍